pkg/collector/collector.go

```diff
@@ -0,0 +1,534 @@
+	helpWorkerTimeout              = "Boolean indicating if the worker upgrade timed out"
+	helpNodeDrainFailed            = "Boolean indicating if a force node drain has failed"
+	helpUpgradeWindowBreached      = "Boolean indicating if a the upgrade window has been breached"
+	helpCollectorFailed            = "An error occurred during scape of metrics"
+
+	// OSD
+	helpNotificationEventSentFailed = "Boolean indicating if a the upgrade notification event has failed"
+)
+
+// CoreUpgradeMetrics holds fields that contain upgrade metrics required to
+// report during an upgrade. These metrics will be consistent regardless of the
+// the upgrade.spec.type.
+type CoreUpgradeMetrics struct {
+	upgradeState              *prometheus.Desc
+	configInvalid             *prometheus.Desc
+	upgradeHealthCheckFailed  *prometheus.Desc
+	scalingFailed             *prometheus.Desc
+	clusterVerificationFailed *prometheus.Desc
+	controlPlaneTimeout       *prometheus.Desc
+	workerTimeout             *prometheus.Desc
+	nodeDrainFailed           *prometheus.Desc
+	upgradeWindowBreached     *prometheus.Desc
+}
+
+// OSDUpgradeMetrics holds metrics specific to the OSD upgrader requirements.
+type OSDUpgradeMetrics struct {
+	// TODO: This is called outside of Reconciles
+	//providerSyncFailed    *prometheus.Desc
+	notificationEventSentFailed *prometheus.Desc
+}
+
+// AROUpgradeMetics holds metrics specific to the OSD upgrader requirements.
+type AROUpgradeMetrics struct {
+}
+
+// UpgradeCollector is implementing prometheus.Collector interface.
+// All these metrics will need to be registered at bootstrap via main.go
+// as we can not all MustRegister every Reconcile.
+// TODO/Idea [dofinn]: Deregister metrics != .spec.type at runtime?
+type UpgradeCollector struct {
+	upgradeConfigManager ucm.UpgradeConfigManager
+	cvClient             cv.ClusterVersion
+	coreUpgradeMetrics   *CoreUpgradeMetrics
+	osdUpgradeMetrics    *OSDUpgradeMetrics
+	aroUpgradeMetrics    *AROUpgradeMetrics
+}
+
+// Consturct a new UpgradeCollector and return to caller.
+func NewUpgradeCollector(c client.Client) (prometheus.Collector, error) {
+	upgradeConfigManager, err := ucm.NewBuilder().NewManager(c)
+	if err != nil {
+		return nil, err
+	}
+
+	coreMetrics := &CoreUpgradeMetrics{
+		upgradeState: prometheus.NewDesc(
+			prometheus.BuildFQName(MetricsNamespace, subSystemUpgrade, "state_timestamp"),
+			helpUpgradeState,
+			[]string{
+				keyVersion,
+				keyPhase,
+			}, nil),
+		configInvalid: prometheus.NewDesc(
+			prometheus.BuildFQName(MetricsNamespace, subSystemUpgradeConfig, "invalid"),
+			helpConfigInvalid,
+			[]string{
+				keyUpgradeConfigName,
+			}, nil),
+		upgradeHealthCheckFailed: prometheus.NewDesc(
+			prometheus.BuildFQName(MetricsNamespace, subSystemCluster, "health_check_failed"),
+			helpUpgradeHealthCheckFailed,
+			[]string{
+				keyUpgradeConfigName,
+				keyState,
+			}, nil),
+		scalingFailed: prometheus.NewDesc(
```

Comment:
Can these upgrader metrics be moved to an osd upgrader collector?

Reply:
This is basically the same thing I'm hoping to see with [my comment](https://github.com/openshift/managed-upgrade-operator/pull/219#discussion_r613004239). Would prefer we establish a clear separation of upgraders early on with this.

Reply:
:+1: Awesome